log4j-slf4j-impl/src/main/java/org/apache/logging/slf4j/Log4jMarkerFactory.java

```diff
@@ -113,26 +125,19 @@ public boolean exists(final String name) {
         return markerMap.containsKey(name);
     }
 
-    /**
-     * Log4j does not support detached Markers. This method always returns false.
-     * @param name The Marker name.
-     * @return {@code false}
-     */
     @Override
     public boolean detachMarker(final String name) {
-        return false;
+        if (name == null) {
+            return false;
+        }
+        detachedMarkers.add(name);
+        return true;
     }
 
-    /**
-     * Log4j does not support detached Markers for performance reasons. The returned Marker is attached.
-     * @param name The Marker name.
-     * @return The named Marker (unmodified).
-     */
     @Override
     public Marker getDetachedMarker(final String name) {
-        LOGGER.warn("Log4j does not support detached Markers. Returned Marker [{}] will be unchanged.", name);
-        return getMarker(name);
+        final org.apache.logging.log4j.Marker log4jMarker = MarkerManager.getMarker(name);
```

Comment:
This `MarkerManager.getMarker(name)` invocation causes the `name` to be tracked in the `org.apache.logging.log4j.MarkerManager.MARKERS` collection, the proposed change only avoids the slf4j adapters cache of markers converted into the slf4j api.

For this to work, log4j2 would also need to be aware of detached markers, which would also require changes to `MarkerManager.Log4jMarker.isInstanceOf(Marker marker)` which currently does a reference equality check rather than checking the string value.

Reply:
@carterkozak thanks for the review! :) 

Yes, this is exactly the point: *I tried to keep Log4j unaware of detached markers*. 

Since Log4j doesn't support detached markers, while SLF4J does, the proposed implementation is wrapping a Lo4j's marker (that is cached internally as you mentioned) with a fresh instance of SLF4J's marker. 

Unless I'm missing something big, there are also few newly written unit tests that can be used to disproof my understanding of the problem.

Reply:
1. I've never understood the point of a detached Marker. A marker is essentially nothing more than a String to be compared. Having multiple instances of the String makes no sense to me (which is why Log4j 2 didn't implement it).
2. My fear with this is that it will lead to a bug report that SLF4J detached markers aren't detached. Saying we don't support them sends a clearer message.

Reply:
thanks @rgoers for your time! In my use case a `Marker` contains health check data, a POJO that gets serialized in JSON for structured logging.  

The only advantage of detached markers is that they can be garbage collected as soon as they not needed anymore... but you're right, this method is not fulfilling this promise (because it is delegating to the `MarkerFactory` that is caching it). 

Let me try to fix that :-)

Reply:
@carterkozak @rgoers I pushed an additional commit to create fully detached markers (just to satisfy the SLF4J's contract). Is that enough? do you see any additional changes? 

Thanks for your time!